pkg/storage/snapshot/restore.go

```diff
@@ -596,6 +606,115 @@ func (t *vmRestoreTarget) reconcileDataVolumes() (bool, error) {
 	return createdDV || waitingDV, nil
 }
 
+func (t *vmRestoreTarget) getControllerRevision(namespace, name string) (*appsv1.ControllerRevision, error) {
+	revisionKey := cacheKeyFunc(namespace, name)
+	obj, exists, err := t.controller.CRInformer.GetStore().GetByKey(revisionKey)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, fmt.Errorf("Unable to find ControllerRevision %s", revisionKey)
+	}
+	return obj.(*appsv1.ControllerRevision), nil
+}
+
+func (t *vmRestoreTarget) getVirtualMachineSnapshot(namespace, name string) (*snapshotv1.VirtualMachineSnapshot, error) {
+	vmSnapshotKey := cacheKeyFunc(namespace, name)
+	obj, exists, err := t.controller.VMSnapshotInformer.GetStore().GetByKey(vmSnapshotKey)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, fmt.Errorf("Unable to find VirtualMachineSnapshot %s", vmSnapshotKey)
+	}
+	return obj.(*snapshotv1.VirtualMachineSnapshot), nil
+}
+
+func (t *vmRestoreTarget) restoreInstancetypeControllerRevision(revisionName, vmSnapshotName string, vm *kubevirtv1.VirtualMachine) (*appsv1.ControllerRevision, error) {
+	existingCR, err := t.getControllerRevision(vm.Namespace, revisionName)
+	if err != nil {
+		return nil, err
+	}
+
+	vmSnapshot, err := t.getVirtualMachineSnapshot(vm.Namespace, vmSnapshotName)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the CR is controlled by the VirtualMachineSnapshot restore it and attempt to claim it if the VirtualMachine exists
+	if metav1.IsControlledBy(existingCR, vmSnapshot) {
+		restoredCR := existingCR.DeepCopy()
+		restoredCR.ObjectMeta.Reset()
+
+		// We remove the snapshot name prefix from the CR and add the name of the restored VM
+		restoredCR.Name = vm.Name + strings.Replace(existingCR.Name, vmSnapshotName, "", 1)
```

Comment:
So IIUC this makes a non-clone restore to point at the original CR (which obviously already exists), right?
So why not call `restoreInstancetypeControllerRevision()` only if it's a cloning restore (i.e. if the target doesn't exist)?

Reply:
That's a valid assumption but I was worried about a corner case where a user has created a VirtualMachineSnapshot, modified the VirtualMachine to say point to a different instancetype CR, maybe deleted the original CR and then wants to restore to the original VirtualMachine. It's definitely a corner case but I think it's still valid to at least attempt to create the CR, if it exists great otherwise we should always get back to the state captured by the VirtualMachineSnapshot.

Reply:
I don't think the scenario you mentioned is possible. Let's say we have a VM and a snapshot. The user can change the VM and replace the instancetype / preference, but he cannot delete the original CR, as the snapshot owns it. In order to delete the CR, the snapshot needs to also be deleted, which makes the restore impossible. IOW, whenever we restore there needs to be a snapshot which owns a CR, which means it necessarily exists.

My main problem with the code right now is that I feel I understand what's going on only because I have a lot of context, and that's also not easy :) I think we should try hard to simplify this as much as we can. WDYT?

Reply:
BTW, what we can do just to completely be on the safe side, is validate (through an informer for example) that the CR truly exists when a non-clone restore occurs. This way we can capture errors in the future if something changes, but would still be able to maintain a code which is straight forward

Reply:
ACK again as discussed I've reworked this.